src/s3/util/s3Tab.ts

```diff
@@ -0,0 +1,103 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+//import { Bucket, DownloadFileRequest, File, S3Client } from '../../shared/clients/s3Client'
+
+//need to subscribe to ondidsave and upload to s3, might need the S3FileNode
+
+export class S3Tab {
+    //private file: File
+    private s3Uri: vscode.Uri
+    private window: typeof vscode.window
+    private editor: vscode.TextEditor | undefined
+
+    //private editing: boolean
+
+    public constructor(private fileUri: vscode.Uri, window = vscode.window) {
+        this.s3Uri = vscode.Uri.parse('s3:' + this.fileUri.fsPath)
+        this.window = window
+    }
+
+    public async openFileOnReadOnly(workspace = vscode.workspace): Promise<vscode.TextEditor | undefined> {
+        //find if there is any active editor for this uri
+        const openEditor = await this.getActiveEditor()
+
+        if (!openEditor) {
+            //there isn't any tab open for this uri, simply open it in read-only with the s3Uri
+            const doc = await workspace.openTextDocument(this.s3Uri)
+            this.editor = await this.window.showTextDocument(doc, { preview: false })
```

Comment:
Yeah so there is a bug when having it open in preview. Let's say a user clicks a 50MB file, and they immediately also open another one of 50B, the latter will get downloaded and displayed first, but due to it is in preview it will get closed and replaced by the 50MB one.

Reply:
I see, good catch. The latest file clicked should always take precedence and we should interrupt any existing files that are still being downloaded. Preview mode feels like the right experience here and otherwise what's the real difference vs. edit mode?

Reply:
What if the user had only +40MB files stored and was trying to review and compare 20 of them? If we cancel each download it would increase the time it takes him to open each file individually by a lot. Otherwise, it can multitask the downloads and one by one will get open on a new tab, if that makes any sense

Reply:
We should discuss this UX further. If this is meant purely as a preview mode I'd expect to quickly flip back and forth between files. If a bunch of new tabs get opened, potentially out of order as well, that is going to get a bit noisy I think. Your point about interrupting downloads is also well taken. Do we surface the progress of downloads to users? For the sake of the preview view itself we should definitely restrict the supported filesize to something pretty small, e.g. 5MB - this can be made configurable if the need arises. Super large files can be opened directly to edit mode, though massive text files are probably going to be rare and generally will slow VSCode itself down quite a bit.

Reply:
Approving this PR as is - let's continue this discussion separately.

Reply:
Can work on this as a P0.5